sympy/codegen/array_utils.py

```diff
@@ -86,6 +86,15 @@ def __new__(cls, expr, *contraction_indices, **kwargs):
         obj._shape = shape
         return obj
 
+    # XXX: Maybe __mul__ and __rmul__ should have some type checking. These
+    # methods were added because Expr no longer accepts non-Expr in __mul__.
```

Comment:
I think that it is possible to avoid calling these methods. This is called from https://github.com/sympy/sympy/blob/270cf6880b441599bda9706e3c8caf696ef20302/sympy/simplify/trigsimp.py#L1170
with `coeff = 1` and `e` an instance of `CoordSys3D` (that appears as the second argument of a `BaseScalar` `B.x`). There could probably be

    return coeff*e if isinstance(e, Expr) else e

Reply:
That's a better idea. I've fixed this in futrig. However I found that I was actually getting the TypeError from multiplying `Tuple` there. I guess that there is some exception masking going on because there is an `except TypeError: pass` in trigsimp - probably that should be removed as it is maybe hiding bugs.

Reply:
All `try ... except` code blocks are potential for trouble. We should discourage developers from using `try ... except`.

Reply:
Agreed. There are good ways to use `try/except` but I see so many bad ways of doing it in SymPy code that I think we should discourage all try/except unless absolutely necessary and even then insist that the scope of try/except be limited as much as possible.